demos/common/processing/include/async_pipeline.h

```diff
@@ -0,0 +1,104 @@
+    /// If it is omitted, new instance of InferenceEngine::Core will be created inside.
+    PipelineBase(std::unique_ptr<ModelBase> modelInstance, const CnnConfig& cnnConfig, InferenceEngine::Core& engine);
+    virtual ~PipelineBase();
+
+    /// Waits until either output data becomes available or pipeline allows to submit more input data.
+    /// Function will treat results as ready only if next sequential result (frame) is ready.
+    void waitForData();
+
+    /// Returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing.
+    /// @returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing, false otherwise.
+    bool isReadyToProcess() { return requestsPool->isIdleRequestAvailable(); }
+
+    /// Returns performance metrics
+    /// @returns performance metrics structure
+    const PerformanceMetrics getMetrics() const { return perfMetrics; }
+
+    /// Waits for all currently submitted requests to be completed.
+    ///
+    void waitForTotalCompletion() { if (requestsPool) requestsPool->waitForTotalCompletion(); }
+
+    /// Submit request to network
+    /// @param image - image to submit for processing
+    /// @returns -1 if image cannot be scheduled for processing (there's no free InferRequest available).
+    /// Otherwise returns unique sequential frame ID for this particular request. Same frame ID will be written in the response structure.
+    virtual int64_t submitImage(cv::Mat img);
+
+    /// Gets available data from the queue 
+    /// Function will treat results as ready only if next sequential result (frame) is ready.
+    virtual std::unique_ptr<ResultBase> getResult();
+
+protected:
+    /// Submit request to network
+    /// @param request - request to be submitted (caller function should obtain it using getIdleRequest)
+    /// @param metaData - additional source data. This is optional transparent data not used in inference process directly.
+    /// It is passed to inference result directly and can be used in postprocessing.
+    /// @returns unique sequential frame ID for this particular request. Same frame ID will be written in the responce structure.
+    virtual int64_t submitRequest(const InferenceEngine::InferRequest::Ptr& request,const std::shared_ptr<MetaData>& metaData);
+
+    /// Returns processed result, if available
+    /// Function will treat results as ready only if next sequential result (frame) is ready.
+    /// @returns InferenceResult with processed information or empty InferenceResult (with negative frameID) if there's no any results yet.
+    virtual InferenceResult getInferenceResult();
+
+protected:
+    std::unique_ptr<RequestsPool> requestsPool;
+    std::unordered_map<int64_t, InferenceResult> completedInferenceResults;
+
+    InferenceEngine::ExecutableNetwork execNetwork;
+
+    PerformanceMetrics perfMetrics;
```

Comment:
That is done for unification and code reduction.  But you're right, metrics update should be done at the very end of getResult (to include postprocessing stage). I've moved it there. 
However, there's no need to move PerformaceMetrics to application code as it covers whole inference process and doesn't include only rendering time (renderDetectionData). But as long as rendering is an application level task, it's OK to remove its processing time from whole latency.

Reply:
In the end when you tweak parameters for presenting, you care about overall latency because it represent a lag between capturing your movement in front of camera and its appearance.

Reply:
In response to https://github.com/openvinotoolkit/open_model_zoo/pull/1639#issuecomment-725329625

> But I still think that we should measure only pipeline latency and not include application level delays into it.

As far as I remember `application level` is a fixed set of steps which you just listed in your definition. That means they don't have any general property we should care about. It's just your understanding how to group steps. You have to come up with good property you want to preserve to make it a valid argument.

> It is better for comparison of different model cases (as camera of stream reader might give unexpected delays).

With time we could come up with an acceptable latency, say 0.2 sec, which we agree is of reasonable value to tweak parameters in a way so latency stays around it. Putting results on images for different demos can take different time while latency should be kept the same.

> Besides, it really simplifies the code – you don’t need to put extra measurements into application code.

Now it looks it makes code more complex. You've split one of the methods on two, you have that mess with includes.

Reply:
done